File: leetcode/0946-validate-stack-sequences/solution.py
from typing import List


class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stack = []
        pop_index = 0

        for val in pushed:
            stack.append(val)
            while stack and pop_index < len(popped) and stack[-1] == popped[pop_index]:
                stack.pop()
                pop_index += 1

        return not stack

File: leetcode/0946-validate-stack-sequences/solution.cpp
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        stack<int> st;
        int popIndex = 0;

        for (int val : pushed) {
            st.push(val);
            while (!st.empty() && popIndex < (int)popped.size() && st.top() == popped[popIndex]) {
                st.pop();
                popIndex++;
            }
        }

        return st.empty();
    }
};
